libraries/chain/hotstuff/finalizer.cpp

```diff
@@ -0,0 +1,302 @@
+#include <fc/log/logger_config.hpp>
+
+namespace eosio::chain {
+
+// ----------------------------------------------------------------------------------------
+block_state_ptr get_block_by_height(const fork_database_if_t::branch_type& branch, uint32_t block_num) {
+   auto it = std::find_if(branch.begin(), branch.end(),
+                          [&](const block_state_ptr& bsp) { return bsp->block_num() == block_num; });
+   return it == branch.end() ? block_state_ptr{} : *it;
+}
+
+// ----------------------------------------------------------------------------------------
+qc_chain_t finalizer::get_qc_chain(const block_state_ptr& proposal, const branch_type& branch) const {
+   qc_chain_t res;
+
+   // get b2
+   // ------
+   auto it2 = std::find_if(branch.begin(), branch.end(),
+        [t=proposal->core.last_qc_block_num](const block_state_ptr& bsp) { return bsp->block_num() == t; });
+   if (it2 == branch.end())
+      return res;
+   res.b2 = *it2;
+
+   // get b1
+   // ------
+   auto it1 = std::find_if(++it2, branch.end(),
+        [t=res.b2->core.last_qc_block_num](const block_state_ptr& bsp) { return bsp->block_num() == t; });
+   if (it1 == branch.end())
+      return res;
+   res.b1 = *it1;
+
+   // get b
+   // ------
+   auto it = std::find_if(++it1, branch.end(),
+        [t=res.b1->core.last_qc_block_num](const block_state_ptr& bsp) { return bsp->block_num() == t; });
+   if (it == branch.end())
+      return res;
+   res.b = *it;
+
+   return res;
+}
+
+// ----------------------------------------------------------------------------------------
+bool extends(const fork_database_if_t& fork_db, const block_state_ptr& descendant, const block_id_type& ancestor) {
+   if (ancestor.empty())
+      return false;
+   auto cur = fork_db.get_block_header(descendant->previous()); // use `get_block_header` so we can get the root
+   while (cur) {
+      if (cur->id() == ancestor)
+         return true;
+      cur = fork_db.get_block_header(cur->previous());
+   }
+   return false;
+}
+
+// ----------------------------------------------------------------------------------------
+finalizer::VoteDecision finalizer::decide_vote(const block_state_ptr& p, const fork_database_if_t& fork_db) {
+   bool safety_check   = false;
+   bool liveness_check = false;
+
+   auto p_branch = fork_db.fetch_branch(p->id());
+
+   qc_chain_t chain = get_qc_chain(p, p_branch);
+
+   // we expect last_qc_block_num() to always be found except at bootstrap
+   // in `assemble_block()`, if we don't find a qc in the ancestors of the proposed block, we use block_num
+   // from fork_db.root(), and specify weak.
+   auto bsp_last_qc = p->last_qc_block_num() ? get_block_by_height(p_branch, *p->last_qc_block_num()) : block_state_ptr{};
+
+   bool monotony_check = !fsi.last_vote || p->timestamp() > fsi.last_vote.timestamp;
+   // !fsi.last_vote means we have never voted on a proposal, so the protocol feature just activated and we can proceed
```

Comment:
Change `proposal` to `block`?

Reply:
I don't know. We have used proposal a lot when talking about hotstuff. We could talk about it at the next meeting and ask Areg if we should use block exclusively instead of proposal. My personal feeling is that, in finality related code, using proposal makes it clear what info from the block we are interested in, and the type makes it clear it is a block we are looking at.

Reply:
OK. Not change it at the moment.